core/trino-spi/src/main/java/io/trino/spi/SplitWeight.java

```diff
@@ -0,0 +1,112 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.spi;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonValue;
+
+import java.math.BigDecimal;
+import java.util.Collection;
+import java.util.function.Function;
+
+import static java.lang.Math.addExact;
+import static java.lang.Math.multiplyExact;
+
+public final class SplitWeight
+{
+    private static final long UNIT_VALUE = 100;
+    private static final int UNIT_SCALE = 2; // Decimal scale such that (10 ^ UNIT_SCALE) == UNIT_VALUE
+    private static final SplitWeight STANDARD_WEIGHT = new SplitWeight(UNIT_VALUE);
+
+    private final long value;
+
+    private SplitWeight(long value)
+    {
+        if (value <= 0) {
+            throw new IllegalArgumentException("value must be > 0, found: " + value);
+        }
+        this.value = value;
+    }
+
+    /**
+     * @return The internal integer representation for this weight value
+     */
+    @JsonValue
+    public long getRawValue()
+    {
+        return value;
+    }
+
+    @Override
+    public boolean equals(Object other)
+    {
+        if (!(other instanceof SplitWeight)) {
+            return false;
+        }
+        return this.value == ((SplitWeight) other).value;
+    }
+
+    @Override
+    public int hashCode()
+    {
+        return Long.hashCode(value);
+    }
+
+    @Override
+    public String toString()
+    {
+        if (value == UNIT_VALUE) {
+            return "1";
+        }
+        return BigDecimal.valueOf(value, -UNIT_SCALE).stripTrailingZeros().toPlainString();
+    }
+
+    @JsonCreator
+    public static SplitWeight fromRawValue(long value)
+    {
+        return value == UNIT_VALUE ? STANDARD_WEIGHT : new SplitWeight(value);
+    }
+
+    public static SplitWeight fromProportion(double weight)
+    {
+        if (weight <= 0 || !Double.isFinite(weight)) {
+            throw new IllegalArgumentException("Invalid weight: " + weight);
+        }
+        // Must round up to avoid small weights rounding to 0
+        return fromRawValue((long) Math.ceil(weight * UNIT_VALUE));
+    }
+
+    public static SplitWeight standard()
+    {
+        return STANDARD_WEIGHT;
+    }
+
+    public static long rawValueForStandardSplitCount(int splitCount)
+    {
+        if (splitCount < 0) {
+            throw new IllegalArgumentException("splitCount must be >= 0, found: " + splitCount);
+        }
+        return multiplyExact(splitCount, UNIT_VALUE);
+    }
+
+    public static <T> long rawValueSum(Collection<T> collection, Function<T, SplitWeight> getter)
+    {
+        long sum = 0;
+        for (T item : collection) {
+            long value = getter.apply(item).getRawValue();
+            sum = addExact(sum, value);
+        }
+        return sum;
+    }
```

Comment:
Is it supposed to be part of SPI? is the connector supposed to invoke these methods? when?

Reply:
These are part of the SPI only incidentally because the engine needs to use these utilities in various components. They could be moved to `trino-main`, but then they'd be duplicated in different parts of the code that may or may not perform the same overflow validation.

Connectors should not need to know or care about the "raw value" representation, and should instead prefer to express weights relative to a "standard weight" which is to say, they are supposed to use `SplitWeight.fromProportion(double proportion)` which takes any value relative to `1.0`. Weights must be positive and non-zero, but otherwise connectors are free to express any value that makes sense for that connector to influence the number of splits that are allowed to be assigned to workers and tasks by the scheduler. Internally, we use integer numeric representation, but that is an implementation detail for the engine which is required in order to avoid accumulating account errors due to precision loss which would occur had we used the `double` representation within the engine itself and in the serialized JSON encoding.

Reply:
> These are part of the SPI only incidentally because the engine needs to use these utilities in various components. They could be moved to `trino-main`, but then they'd be duplicated in different parts of the code that may or may not perform the same overflow validation.

Let's move them to a shared utility class to avoid duplication

> Connectors should not need to know or care about the "raw value" representation, and should instead prefer to express weights relative to a "standard weight" which is to say, they are supposed to use SplitWeight.fromProportion(double proportion) which takes any value relative to 1.0. 

Can you please add some javadoc? In particular, it's unclear "proportion  to what" this is, and what's the range of allowed values.

Reply:
javadoc and scheduler configuration documentation added in https://github.com/trinodb/trino/pull/12684